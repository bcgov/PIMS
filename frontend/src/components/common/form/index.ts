export * from './Button';
export * from './Check';
export * from './DisplayCurrency';
export * from './DisplayError';
export * from './FastCurrencyInput';
export * from './FastDatePicker';
export * from './FastFiscalYearInput';
export * from './FastInput';
export * from './FastSelect';
export * from './Field';
export * from './Form';
export * from './Input';
export * from './InputGroup';
export * from './ParentSelect';
export * from './Select';
export * from './TextArea';
