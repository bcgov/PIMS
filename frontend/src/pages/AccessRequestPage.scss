.userInfo {
  width: 40%;
  vertical-align: middle;
  text-align: left;
  display: table;
}

.success {
  display: flex;
  background-color: lightgreen;
  margin: auto;
  padding: 5px;
}

.success p {
  text-align: center;
  vertical-align: middle;
  margin: auto;
}

.info {
  display: flex;
  background-color: lightskyblue;
  margin: auto;
  padding: 5px;
}

.info p {
  text-align: center;
  vertical-align: middle;
  margin: auto;
}

.notes {
  padding-bottom: 30px;
}

.isDisabled {
  padding-bottom: 10px;
}

.cancelSave {
  padding-top: 15px;
  padding-bottom: 10px;
}

.navBar {
  font-weight: 700;
  text-align: left;
}

div.required {
  color: black;
  font-weight: normal;
}

.required {
  color: red;
}
