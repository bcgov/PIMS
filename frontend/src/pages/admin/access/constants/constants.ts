import { ColDef } from 'ag-grid-community';
import { IAccessRequest } from 'interfaces';
import { AccessStatusDisplay, AccessStatusDisplayMapper } from 'constants/accessStatus';

export const columnDefinitions = (): ColDef[] => {
  let items: ColDef[] = [
    {
      headerName: 'IDIR/BCeID',
      checkboxSelection: true,
      headerCheckboxSelection: true,
      valueGetter: params => (params.data as IAccessRequest).user.username,
    },
    {
      headerName: 'First name',
      valueGetter: params => (params.data as IAccessRequest).user.firstName,
    },
    {
      headerName: 'Last name',
      valueGetter: params => (params.data as IAccessRequest).user.lastName,
    },
    {
      headerName: 'Email',
      valueGetter: params => (params.data as IAccessRequest).user.email,
      width: 250,
    },
    {
      headerName: 'Position',
      valueGetter: params => (params.data as IAccessRequest).position,
    },
    {
      headerName: 'Status',
      valueGetter: (params): AccessStatusDisplay =>
        AccessStatusDisplayMapper[(params.data as IAccessRequest).status],
      filter: false,
      sortable: false,
    },
    {
      headerName: 'Agency',
      valueGetter: params => (params.data as IAccessRequest).agencies[0].name,
      width: 250,
    },
    {
      headerName: 'Role',
      valueGetter: params => (params.data as IAccessRequest).roles[0].name,
    },
  ];

  items = items.map(column => ({
    filter: true,
    suppressMenu: true,
    cellClass: 'user-info-cell',
    sortable: true,
    minWidth: 230,
    ...column,
  }));

  return items;
};
