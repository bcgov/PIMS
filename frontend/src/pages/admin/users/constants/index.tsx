import * as React from 'react';
import { MUIDataTableColumnDef } from 'mui-datatables';
import { FaCheck, FaTimes } from 'react-icons/fa';
import { AccountActive } from '../interfaces/IUserRecord';
import { RowActions } from '../components/RowActions';
export const columns: MUIDataTableColumnDef[] = [
  {
    name: 'id',
    label: 'ID',
    options: {
      display: 'excluded',
      filter: false,
    },
  },
  {
    name: 'username',
    label: 'IDIR/BCeID',
    options: {
      customBodyRender: (username: string, tableMeta: any) => {
        return <a href={`/admin/user/${tableMeta.rowData[0]}`}>{username}</a>;
      },
    },
  },
  {
    name: 'firstName',
    label: 'First name',
  },
  {
    name: 'lastName',
    label: 'Last name',
  },
  {
    name: 'email',
    label: 'Email',
  },
  {
    name: 'active',
    label: 'Active',
    options: {
      customBodyRender: (active: AccountActive, tableMeta: any) => {
        return active === AccountActive.YES ? <FaCheck /> : <FaTimes />;
      },
    },
  },
  {
    name: 'agency',
    label: 'Agency',
  },
  {
    name: 'position',
    label: 'Position',
  },
  {
    name: 'role',
    label: 'Role',
  },
  {
    name: 'action',
    label: 'Actions',
    options: {
      customBodyRender: (value: any, tableMeta: any) => {
        return <RowActions userId={tableMeta.rowData[0]} active={tableMeta.rowData[5]} />;
      },
    },
  },
];
