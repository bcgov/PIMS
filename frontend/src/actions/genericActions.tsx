import * as ActionTypes from 'constants/actionTypes';

export interface IGenericNetworkAction {
  isFetching: boolean;
  name: string;
  type: string;
  errorMessage?: string;
  status?: number;
  data?: any;
}

export const success = (reducer: string, status?: number, data?: any): IGenericNetworkAction => ({
  isFetching: false,
  name: reducer,
  type: ActionTypes.SUCCESS,
  status,
  data,
});

export const request = (reducer: string): IGenericNetworkAction => ({
  isFetching: true,
  name: reducer,
  type: ActionTypes.REQUEST,
});

export const error = (reducer: string, status?: number, err?: any): IGenericNetworkAction => ({
  isFetching: false,
  name: reducer,
  type: ActionTypes.ERROR,
  errorMessage: err,
  status,
});

export const clear = (reducer: string): IGenericNetworkAction => ({
  isFetching: false,
  name: reducer,
  type: ActionTypes.CLEAR,
});
