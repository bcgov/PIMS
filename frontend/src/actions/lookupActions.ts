import * as ActionTypes from 'constants/actionTypes';

//Parcel List API action

export interface ILookupCode {
  code: string;
  name: string;
  id: string;
  isDisabled: boolean;
  isPublic?: boolean;
  type: string;
  parentId?: number;
}

export interface IStoreLookupCodeAction {
  type: typeof ActionTypes.STORE_LOOKUP_CODE_RESULTS;
  lookupCodes: ILookupCode[];
}

export const storeLookupCodesAction = (lookupCodes: ILookupCode[]) => ({
  type: ActionTypes.STORE_LOOKUP_CODE_RESULTS,
  lookupCodes: lookupCodes,
});
