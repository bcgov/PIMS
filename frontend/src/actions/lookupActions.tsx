import * as ActionTypes from 'constants/actionTypes';

//Parcel List API action

export interface ILookupCode {
  code: string;
  name: string;
  id: string;
  isDisabled: boolean;
  type: string;
}

export interface IStoreLookupCodeAction {
  type: typeof ActionTypes.STORE_LOOKUP_CODE_RESULTS;
  lookupCodes: ILookupCode[];
}

export const storeLookupCodesAction = (lookupCodes: ILookupCode[]) => ({
  type: ActionTypes.STORE_LOOKUP_CODE_RESULTS,
  lookupCodes: lookupCodes,
});
