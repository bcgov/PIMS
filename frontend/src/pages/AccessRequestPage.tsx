import React, { useEffect } from 'react';
import './AccessRequestPage.scss';
import { Container, Row, Col, ButtonToolbar, Button, Alert } from 'react-bootstrap';
import { IGenericNetworkAction } from 'actions/genericActions';
import { ILookupCode } from 'actions/lookupActions';
import {
  getCurrentAccessRequestAction,
  getSubmitAccessRequestAction,
  toAccessRequest,
} from 'actionCreators/accessRequestActionCreator';
import { IUserInfo, IAccessRequest } from 'interfaces';
import { Formik } from 'formik';
import { Form, Input, TextArea, Select } from '../components/common/form';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from 'reducers/rootReducer';
import { IAccessRequestState } from 'reducers/accessRequestReducer';
import { ILookupCodeState } from 'reducers/lookupCodeReducer';
import * as API from 'constants/API';
import * as actionTypes from 'constants/actionTypes';
import { DISCLAIMER_URL, PRIVACY_POLICY_URL } from 'constants/strings';
import _ from 'lodash';
import { AccessRequestSchema } from 'utils/YupSchema';
import useKeycloakWrapper from 'hooks/useKeycloakWrapper';
import { mapLookupCode } from 'utils';

interface IAccessRequestForm extends IAccessRequest {
  agency: number;
  role: string;
}

/**
 * The AccessRequestPage provides a way to new authenticated users to submit a request
 * that associates them with a specific agency and a role within the agency.
 * If they have an active access request already submitted, it will allow them to update it until it has been approved or disabled.
 * If their prior request was disabled they will then be able to submit a new request.
 */
const AccessRequestPage = () => {
  const keycloakWrapper = useKeycloakWrapper();
  const keycloak = keycloakWrapper.obj;
  const userInfo = keycloak?.userInfo as IUserInfo;
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getCurrentAccessRequestAction());
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const data = useSelector<RootState, IAccessRequestState>(
    state => state.accessRequest as IAccessRequestState,
  );
  const lookupCodes = useSelector<RootState, ILookupCode[]>(
    state => (state.lookupCode as ILookupCodeState).lookupCodes,
  );
  const request = useSelector<RootState, IGenericNetworkAction>(
    state => (state.network as any)[actionTypes.ADD_REQUEST_ACCESS] as IGenericNetworkAction,
  );
  const agencies = _.filter(lookupCodes, (lookupCode: ILookupCode) => {
    return lookupCode.type === API.AGENCY_CODE_SET_NAME;
  });
  const roles = _.filter(lookupCodes, (lookupCode: ILookupCode) => {
    return lookupCode.type === API.ROLE_CODE_SET_NAME;
  });

  const accessRequest = data?.accessRequest;
  const initialValues: IAccessRequestForm = {
    id: accessRequest?.id ?? 0,
    userId: userInfo?.sub,
    user: {
      id: userInfo?.sub,
      username: userInfo?.username,
      displayName: userInfo?.name,
      firstName: userInfo?.firstName,
      lastName: userInfo?.lastName,
      email: userInfo?.email,
      position: accessRequest?.user?.position ?? userInfo?.position ?? '',
    },
    agencies: accessRequest?.agencies ?? [],
    isGranted: accessRequest?.isGranted ?? false,
    roles: accessRequest?.roles ?? [],
    note: accessRequest?.note ?? '',
    agency: accessRequest?.agencies?.find(x => x).id,
    role: accessRequest?.roles?.find(x => x).id,
    rowVersion: accessRequest?.rowVersion,
  };

  const selectAgencies = agencies.map(c => mapLookupCode(c, initialValues.agency));
  const selectRoles = roles.map(c => mapLookupCode(c, initialValues.role));

  const checkAgencies = (
    <Select
      label="Agency"
      field="agency"
      required={true}
      options={selectAgencies}
      placeholder={initialValues?.agencies?.length > 0 ? undefined : 'Please Select'}
    />
  );

  const checkRoles = (
    <Select
      label="Role"
      field="role"
      required={true}
      options={selectRoles}
      placeholder={initialValues?.roles?.length > 0 ? undefined : 'Please Select'}
    />
  );

  const button = initialValues.id === 0 ? 'Submit' : 'Update';
  const inProgress =
    initialValues.id !== 0 ? (
      <Alert key={initialValues.id} variant="primary">
        You will receive an email when your request is reviewed.
      </Alert>
    ) : null;
  const success =
    (!request?.isFetching && request?.status === 201) || request?.status === 200 ? (
      <Alert key={initialValues.id} variant="success">
        Your request has been submitted.
      </Alert>
    ) : null;

  return (
    <div>
      <div>
        <h3>Access Request</h3>
      </div>
      <hr />
      <Container fluid={true}>
        <Row className="justify-content-md-center">
          <Formik
            enableReinitialize={true}
            initialValues={initialValues}
            validationSchema={AccessRequestSchema}
            onSubmit={(values, { setSubmitting }) => {
              dispatch(getSubmitAccessRequestAction(toAccessRequest(values)));
              setSubmitting(false);
            }}
          >
            {props => (
              <Form className="userInfo">
                {success}

                {inProgress}

                <Input
                  label="IDIR/BCeID"
                  field="user.username"
                  placeholder={initialValues.user.username}
                  readOnly={true}
                  type="text"
                />

                <Row>
                  <Col>
                    <Input
                      label="First Name"
                      field="user.firstName"
                      placeholder={initialValues.user.firstName}
                      readOnly={true}
                      type="text"
                    />
                  </Col>
                  <Col>
                    <Input
                      label="Last Name"
                      field="user.lastName"
                      placeholder={initialValues.user.lastName}
                      readOnly={true}
                      type="text"
                    />
                  </Col>
                </Row>

                <Input
                  label="Email"
                  field="user.email"
                  placeholder={initialValues.user.email}
                  readOnly={true}
                  type="email"
                />

                {checkAgencies}

                <Input
                  label="Position"
                  field="user.position"
                  placeholder="e.g) Director, Real Estate and Stakeholder Engagement"
                  type="text"
                />

                {checkRoles}

                <TextArea
                  label="Notes"
                  field="note"
                  placeholder="Please specify why you need access to this site."
                />

                <p>
                  By clicking request, you agree to our{' '}
                  <a href={DISCLAIMER_URL}>Terms and Conditions</a> and that you have read our{' '}
                  <a href={PRIVACY_POLICY_URL}>Privacy Policy</a>.
                </p>

                <Row className="justify-content-md-center">
                  <ButtonToolbar className="cancelSave">
                    <Button className="mr-5" type="submit">
                      {button}
                    </Button>
                  </ButtonToolbar>
                </Row>
              </Form>
            )}
          </Formik>
        </Row>
      </Container>
    </div>
  );
};

export default AccessRequestPage;
