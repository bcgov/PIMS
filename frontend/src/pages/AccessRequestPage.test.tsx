import React from 'react';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme from 'enzyme';
import { Select } from '../components/common/form';
import { Formik } from 'formik';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import AccessRequestPage from './AccessRequestPage';
import { ILookupCode } from '../actions/lookupActions';
import { IGenericNetworkAction } from '../actions/genericActions';
import { NETWORK } from '../constants/reducerTypes';
import * as actionTypes from '../constants/actionTypes';

Enzyme.configure({ adapter: new Adapter() });

const lCodes = [
  { name: 'One', id: '1', isDisabled: false, type: 'core operational' },
] as ILookupCode[];

const requestAccess = {
  status: 200,
  isFetching: false,
} as IGenericNetworkAction;

const mockStore = configureMockStore([thunk]);
const history = createMemoryHistory();

// Simulating a succesful submit
const successStore = mockStore({
  lookupCode: lCodes,
  [NETWORK]: {
    [actionTypes.ADD_REQUEST_ACCESS]: requestAccess,
  },
});

// Store without status of 200
const store = mockStore({
  lookupCode: lCodes,
  [NETWORK]: {
    addRequestAccess: requestAccess,
  },
});

it('renders RequestAccessPage correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Router history={history}>
          <AccessRequestPage />
        </Router>
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe('component functionality when requestAccess status is 200 and fetching is false', () => {
  const componentRender = mount(
    <Provider store={successStore}>
      <Router history={history}>
        <AccessRequestPage />
      </Router>
    </Provider>,
  );

  it('renders dropdown for agenices and roles', () => {
    expect(componentRender.find(Select).length).toBe(2);
  });

  it('initializes form with null for agencies and roles', () => {
    expect(
      componentRender
        .find(Formik)
        .first()
        .prop('initialValues'),
    ).toEqual({
      agencies: [],
      agency: undefined,
      id: 0,
      status: 'OnHold',
      note: '',
      role: undefined,
      roles: [],
      rowVersion: undefined,
      user: {
        displayName: undefined,
        email: undefined,
        firstName: undefined,
        id: undefined,
        lastName: undefined,
        position: '',
        username: undefined,
      },
      userId: undefined,
    });
  });

  it('displays a success message', () => {
    expect(
      componentRender
        .find('div.alert')
        .first()
        .text(),
    ).toContain('Your request has been submitted.');
  });
});

it('does not show success message by default', () => {
  const component = mount(
    <Provider store={store}>
      <Router history={history}>
        <AccessRequestPage />
      </Router>
    </Provider>,
  );
  expect(
    component
      .find('div.alert')
      .first()
      .text(),
  ).toContain('Your request has been submitted.');
});
