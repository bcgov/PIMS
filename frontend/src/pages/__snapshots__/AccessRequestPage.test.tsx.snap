// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders RequestAccessPage correctly 1`] = `
<div>
  <div>
    <h3>
      Access Request
    </h3>
  </div>
  <hr />
  <div
    className="container-fluid"
  >
    <div
      className="justify-content-md-center row"
    >
      <form
        className="userInfo"
        noValidate={true}
        onSubmit={[Function]}
      >
        <div
          className="fade alert alert-success show"
          role="alert"
        >
          Your request has been submitted.
        </div>
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="input-user.username"
          >
            IDIR/BCeID
          </label>
          <input
            className="form-control"
            id="input-user.username"
            name="user.username"
            onChange={[Function]}
            readOnly={true}
            type="text"
          />
        </div>
        <div
          className="row"
        >
          <div
            className="col"
          >
            <div
              className="form-group"
            >
              <label
                className="form-label"
                htmlFor="input-user.firstName"
              >
                First Name
              </label>
              <input
                className="form-control"
                id="input-user.firstName"
                name="user.firstName"
                onChange={[Function]}
                readOnly={true}
                type="text"
              />
            </div>
          </div>
          <div
            className="col"
          >
            <div
              className="form-group"
            >
              <label
                className="form-label"
                htmlFor="input-user.lastName"
              >
                Last Name
              </label>
              <input
                className="form-control"
                id="input-user.lastName"
                name="user.lastName"
                onChange={[Function]}
                readOnly={true}
                type="text"
              />
            </div>
          </div>
        </div>
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="input-user.email"
          >
            Email
          </label>
          <input
            className="form-control"
            id="input-user.email"
            name="user.email"
            onChange={[Function]}
            readOnly={true}
            type="email"
          />
        </div>
        <div
          className="required form-group"
        >
          <label
            className="form-label"
            htmlFor="input-agency"
          >
            Agency
          </label>
          <span
            className="required"
          >
            *
          </span>
          <select
            className="form-control"
            id="input-agency"
            name="agency"
            onChange={[Function]}
            required={true}
          >
            <option
              value=""
            >
              Please Select
            </option>
          </select>
        </div>
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="input-user.position"
          >
            Position
          </label>
          <input
            className="form-control"
            id="input-user.position"
            name="user.position"
            onChange={[Function]}
            placeholder="e.g) Director, Real Estate and Stakeholder Engagement"
            type="text"
            value=""
          />
        </div>
        <div
          className="required form-group"
        >
          <label
            className="form-label"
            htmlFor="input-role"
          >
            Role
          </label>
          <span
            className="required"
          >
            *
          </span>
          <select
            className="form-control"
            id="input-role"
            name="role"
            onChange={[Function]}
            required={true}
          >
            <option
              value=""
            >
              Please Select
            </option>
          </select>
        </div>
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="input-note"
          >
            Notes
          </label>
          <textarea
            className="form-control"
            id="input-note"
            name="note"
            onChange={[Function]}
            placeholder="Please specify why you need access to this site."
            value=""
          />
        </div>
        <p>
          By clicking request, you agree to our
           
          <a
            href="https://www2.gov.bc.ca//gov/content/home/disclaimer"
          >
            Terms and Conditions
          </a>
           and that you have read our
           
          <a
            href="http://www.gov.bc.ca/gov/content/home/privacy"
          >
            Privacy Policy
          </a>
          .
        </p>
        <div
          className="justify-content-md-center row"
        >
          <div
            className="cancelSave btn-toolbar"
            role="toolbar"
          >
            <button
              className="mr-5 btn btn-primary"
              disabled={false}
              type="submit"
            >
              Submit
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
